Change light brightness, control light groups, create user

roguh opened the pull request:
- Add ability to change light brightness
- Add http.post and bridge.create_user
- Add ability to query and control light groups

roguh committed:
Add ability to change light brightness

roguh committed:
Add http.post and bridge.create_user

roguh committed:
Add ability to query and control light groups

roguh committed:
Merge branch 'main' of github.com:roguh/hue-api

roguh commented:
Hello, would you like to accept these changes? I can also split them up into multiple PRs. Any feedback is appreciated. I can make any changes you need.

roguh commented:
There's almost certainly a more efficient way of toggling on all lights a group. I can research this more before you consider this PR, if you want.

nirantak commented:
> Hello, would you like to accept these changes? I can also split them up into multiple PRs. Any feedback is appreciated. I can make any changes you need.

Hi @roguh, I will be happy to accept any changes to this project. It might take a couple of days to review this PR, but in the meanwhile can you please add tests for these changes and fix the Linter check?

nirantak commented:
> There's almost certainly a more efficient way of toggling on all lights a group. I can research this more before you consider this PR, if you want.

It would be great if you could do this

roguh commented:
I'll add tests, run the linter locally, and find a better way to control a light group.

roguh committed:
Remove unused import

roguh committed:
Rename light.brightness -> light.set_brightness

roguh committed:
Save brightness in self.brightness

roguh committed:
Do not send "on" in every call to set_state

roguh committed:
Print coverage stats to terminal, as well as html

roguh committed:
Test brightness.

Coverage for hue/api/light.py and hue/cli/light.py now 100%

roguh committed:
Fix output for brightness

roguh committed:
Add docs to hue bridge create-user command

roguh committed:
Test hue bridge create user

Coverage in hue/api/bridge.py and hue/cli/bridge.py is 100%

roguh committed:
Test hue.api.http.post

Coverage is now 100% in hue/api/http

roguh committed:
Optimize light group control with specialized path. Add on, off, and brightness to light groups

roguh commented:
I've added tests and fixed the light group control so it is far more efficient and reliable.

I still need to add tests to the light group code.

The only code of yours that I've changed is where you sent `"on"` in every call to set_state. I think it's better to only send `"on"` when turning a light on or off.

nirantak commented on hue/api/light.py:
```diff
@@ -153,3 +157,13 @@ async def toggle(self) -> list[dict[str, dict[str, Any]]]:
+        Returns:
+            A list of dictionaries with key=success/error and value=state element changed
+        """
+        await self.get_state()
```
```suggestion
```
This line is not needed

nirantak replied:
Also, please do update the docstring for description and arguments

nirantak commented on hue/api/light_group.py:
```diff
@@ -0,0 +1,103 @@
+
+class LightGroup(Bridge):
+    """
+    Interact with the Hue Lights API
+
+    Example:
+        >>> from hue import LightGroup
+        >>> bridge = LightGroup("All", ip="192.168.1.10", user="xxxx")
+
+    Attributes:
+        name: Name of the Hue Light group
```
```suggestion
    Interact with the Hue Light Groups API

    Example:
        >>> from hue import LightGroup
        >>> group = LightGroup("All", ip="192.168.1.10", user="xxxx")

    Attributes:
        name: Name of the Light Group
```

nirantak commented on hue/api/light_group.py:
```diff
@@ -0,0 +1,103 @@
+
+    def __init__(self: LightGroup, name: str, *, ip: str, user: str) -> LightGroup:
+        """
+        Initialize a Hue Light object
+
+        Arguments:
+            name: The name of the Light group to control
+            ip: IP address of the Hue Bridge, get it by running `hue bridge discover`
+            user: The secret user id created to access the Hue API
+
+        Returns:
+            An object of class Light
```
```suggestion
        Initialize a Hue Light Group object

        Arguments:
            name: The name of the Light Group to control
            ip: IP address of the Hue Bridge, get it by running `hue bridge discover`
            user: The secret user id created to access the Hue API

        Returns:
            An object of class LightGroup
```

nirantak commented on hue/api/light_group.py:
```diff
@@ -0,0 +1,103 @@
+            print(exc)
+        return None
+
+    async def set_state(self, state):
```
Can you please add type hints for all method arguments and return values?

nirantak commented on hue/cli/bridge.py:
```diff
@@ -36,3 +36,16 @@ def get(
+    ip: str = typer.Option(..., "--ip", "-i", envvar="HUE_BRIDGE_IP"),
+    device_type: str = typer.Option(..., "--device-type", "-d"),
+):
+    """Create a user on a Hue Bridge.
+
+    You must press the button on the Bridge.
```
```suggestion
    """
    Create a user for the Hue Bridge.
    You must press the button on the Bridge atmost 30 seconds before running this command.
```

nirantak commented on hue/cli/bridge.py:
```diff
@@ -36,3 +36,16 @@ def get(
+    You must press the button on the Bridge.
+    """
+    resp = asyncio.run(Bridge.create_user(ip, device_type))
+    console.print(f"[{ip}] Bridge Config:\n{json.dumps(resp, indent=2)}")
```
```suggestion
    console.print(f"[{ip}] User Config:\n{json.dumps(resp, indent=2)}")
```

nirantak commented on hue/cli/lights.py:
```diff
@@ -0,0 +1,80 @@
+import asyncio
```
Please rename this file to light_group.py to keep the same pattern as the API directory

nirantak commented:
Hi @roguh, let me know if you plan to work on this, or I can update and start adding these in bits and pieces.

roguh commented:
My apologies, I haven't found time for this. You can continue this work, if you wish.

podulator commented:
Hey, any chance of this getting merged at all, it would be super useful for a pet project :)
